include/deal.II/base/vectorization.h

```diff
@@ -3941,6 +3941,211 @@ operator<<(std::ostream &out, const VectorizedArray<Number, width> &p)
   return out;
 }
 
+
+/**
+ * @name Ternary operations on VectorizedArray
+ */
+//@{
+
+/**
+ * TODO
+ */
+enum SIMDComparison : int
+{
+  equal                 = _CMP_EQ_OQ,
+  not_equal             = _CMP_NEQ_OQ,
+  less_than             = _CMP_LT_OQ,
+  less_than_or_equal    = _CMP_LE_OQ,
+  greater_than          = _CMP_GT_OQ,
+  greater_than_or_equal = _CMP_GE_OQ
+};
+
+
+/**
+ * TODO
+ */
+template <SIMDComparison predicate, typename Number>
+DEAL_II_ALWAYS_INLINE inline Number
+compare_and_apply_mask(const Number &left,
+                       const Number &right,
+                       const Number &true_value,
+                       const Number &false_value)
+{
+  bool mask;
+
+  switch (predicate)
+    {
+      case SIMDComparison::equal:
+        mask = (left == right);
+        break;
+      case SIMDComparison::not_equal:
+        mask = (left != right);
+        break;
+      case SIMDComparison::less_than:
+        mask = (left < right);
+        break;
+      case SIMDComparison::less_than_or_equal:
+        mask = (left <= right);
+        break;
+      case SIMDComparison::greater_than:
+        mask = (left > right);
+        break;
+      case SIMDComparison::greater_than_or_equal:
+        mask = (left >= right);
+        break;
+    }
+
+  return mask ? true_value : false_value;
+}
+
+
+/**
+ * TODO
+ */
+template <SIMDComparison predicate, typename Number>
+DEAL_II_ALWAYS_INLINE inline VectorizedArray<Number, 1>
+compare_and_apply_mask(const VectorizedArray<Number, 1> &left,
+                       const VectorizedArray<Number, 1> &right,
+                       const VectorizedArray<Number, 1> &true_value,
+                       const VectorizedArray<Number, 1> &false_value)
+{
+  bool mask;
+
+  switch (predicate)
+    {
+      case SIMDComparison::equal:
+        mask = (left.data == right.data);
+        break;
+      case SIMDComparison::not_equal:
+        mask = (left.data != right.data);
+        break;
+      case SIMDComparison::less_than:
+        mask = (left.data < right.data);
+        break;
+      case SIMDComparison::less_than_or_equal:
+        mask = (left.data <= right.data);
+        break;
+      case SIMDComparison::greater_than:
+        mask = (left.data > right.data);
+        break;
+      case SIMDComparison::greater_than_or_equal:
+        mask = (left.data >= right.data);
+        break;
+    }
+
+  return mask ? true_value.data : false_value.data;
+}
+
+//@}
+
+#ifndef DOXYGEN
+#  if DEAL_II_COMPILER_VECTORIZATION_LEVEL >= 3 && defined(__AVX512F__)
+
+template <int predicate>
+DEAL_II_ALWAYS_INLINE inline VectorizedArray<float, 16>
+compare_and_apply_mask(const VectorizedArray<float, 16> &left,
+                       const VectorizedArray<float, 16> &right,
+                       const VectorizedArray<float, 16> &true_values,
+                       const VectorizedArray<float, 16> &false_values)
+{
+  const auto mask = _mm512_cmp_ps(left.data, right.data, predicate);
```

Comment:
I must admit that I had to check up the type behind `auto`, and found that it is not as easy in the AVX512 case. It seems Intel changed the name to `_mm512_cmp_ps_mask` and returns `__mmask16`. The next difficulty is that the `_mm512_and_ps` and `_mm512_or_ps` functions are not in AVX512-F, the common set available in all supported machines, but only in the AVX512-DQ. Market segmentation of Intel - so annoying.

Reply:
I will look into how to implement those later, I guess we can use some functions operating on the masks directly.